eyeball-im-util/src/vector/limit.rs

```diff
@@ -0,0 +1,367 @@
+    L: Stream<Item = usize>,
+{
+    type Item = S::Item;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut this = self.project();
+        this.poll_next(cx)
+    }
+}
+
+impl<S, L> DynamicLimitProj<'_, S, L>
+where
+    S: Stream,
+    S::Item: VectorDiffContainer,
+    VectorDiffContainerStreamElement<S>: Clone + Send + Sync + 'static,
+    L: Stream<Item = usize>,
+{
+    fn poll_next(&mut self, cx: &mut task::Context<'_>) -> Poll<Option<S::Item>> {
+        // First off, if any value is ready, let's return it.
+        if let Some(ready_value) = self.ready_values.pop_front() {
+            return Poll::Ready(Some(ready_value));
+        }
+
+        // Let's poll a new limit from `limit_stream` before polling `inner_stream`.
+        let diffs_after_limit_update = match self.limit_stream.as_mut().poll_next(cx) {
+            // There is a new limit!
+            Poll::Ready(Some(next_limit)) => self.update_limit(next_limit),
+
+            // Stream has stopped. Let's stopped this one.
+            Poll::Ready(None) => return Poll::Ready(None),
+
+            // No new limit. That's OK.
+            Poll::Pending => None,
+        };
+
+        // We have new `VectorDiff` after the limit has been updated. Let's
+        // return them.
+        if let Some(diffs) = diffs_after_limit_update {
+            return Poll::Ready(Some(diffs));
+        }
+
+        // Now, let's poll `VectorDiff`s from the `inner_stream`.
+        let Some(diffs) = ready!(self.inner_stream.as_mut().poll_next(cx)) else {
+            return Poll::Ready(None);
+        };
+
+        // Now, let's consume and apply the diffs if possible.
+        diffs.for_each(|diff| self.apply_diff(diff));
+
+        // If any value is ready, let's return it.
+        match self.ready_values.pop_front() {
+            Some(diff) => Poll::Ready(Some(diff)),
+            None => Poll::Pending,
+        }
+    }
+
+    fn apply_diff(&mut self, diff: VectorDiff<VectorDiffContainerStreamElement<S>>) {
+        let limit = *self.limit;
+        let length = self.buffered_vector.len();
+
+        // Let's update the `buffered_vector`. It's a replica of the original observed
+        // `Vector`. We need to maintain it in order to be able to produce valid
+        // `VectorDiff` when items are missing.
+        self.update_buffered_vector(&diff);
+
+        // If the limit, we have nothing to do.
+        if limit == 0 {
+            return;
+        }
+
+        let is_full = length >= limit;
+        let has_values_after_limit = length > limit;
+
+        match diff {
+            VectorDiff::Append { mut values } => {
+                if is_full {
+                    // Let's ignore the diff.
+                } else {
+                    // Let's truncate the `values` to fit inside the free space.
+                    values.truncate(min(limit - length, values.len()));
```

Comment:
Imagine `limit = 4`, `length = 0` (there is zero items in the buffered vector), and `values.len() = 3` (3 items are being pushed). `values.truncate(limit - length)` will raise a panic from `Vector::split_off` because `split_index > vector.len()`: we will truncate at 4 - 0 = 4, but it contains only 3 items.

Basically, this `min` is here to avoid truncating at `index > values.len()`.

Reply:
Oh right, it's `Vector::truncate`, not `ObservableVector::truncate`. So basically we need https://github.com/jneem/imbl/pull/67 to be released.